clib/hashcons.ml

```diff
@@ -65,8 +65,11 @@ module Make (X : HashconsedType) : (S with type t = X.t and type u = X.u) =
       (tab, u)
 
     let hcons (tab, u) x =
-      let y = X.hashcons u x in
-      Htbl.repr (X.hash y) y tab
+      let h = X.hash x in
+      if Htbl.mem h x tab then x
+      else
+        let y = X.hashcons u x in
+        Htbl.repr h y tab
```

Comment:
Two questions:
1. We are performing a total of two lookups when Htbl.mem returns false. Could that be part of why this PR can be slower than master?
2. If Htbl.mem succeeds then you return `x` whereas the original code would have returned the representative of `x` in the hashmap. Wouldn't this lead to less sharing?

Reply:
Oh, I think my second question doesn't make sense. Your implementation of `mem` actually checks for physical equality so returning `x` is also what would have happened in the previous version of this code.

Reply:
To answer your first question, maybe but I am not sure. I've been conducing local experiments and I don't know what to think. I'll retry with a refined version of this PR and see what happens.